pants-plugins/api_spec/rules.py

```diff
@@ -0,0 +1,259 @@
+                "--config-file",
+                "conf/st2.dev.conf",
+            ),
+            input_digest=input_digest,
+            description="Regenerating openapi.yaml api spec",
+            level=LogLevel.DEBUG,
+        ),
+    )
+
+    contents = [
+        FileContent(
+            f"{field_set.address.spec_path}/{field_set.source.value}",
+            result.stdout,
+        )
+        for field_set in request.field_sets
+    ]
+
+    output_digest = await Get(Digest, CreateDigest(contents))
+    output_snapshot = await Get(Snapshot, Digest, output_digest)
+    # TODO: Drop result.stdout since we already wrote it to a file?
+    return FmtResult.create(request, result, output_snapshot, strip_chroot_path=True)
+
+
+@rule(
+    desc="Validate openapi.yaml with st2-validate-api-spec",
+    level=LogLevel.DEBUG,
+)
+async def validate_api_spec(
+    request: ValidateAPISpecRequest,
+) -> LintResults:
+    # There will only be one target+field_set, but we iterate
+    # to satisfy how lint expects that there could be more than one.
+    # If there is more than one, they will all get the same contents.
+
+    # Find all the dependencies of our target
+    transitive_targets = await Get(
+        TransitiveTargets,
+        TransitiveTargetsRequest(
+            [field_set.address for field_set in request.field_sets]
+        ),
+    )
+
+    dependency_files_get = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[
+                tgt.get(SourcesField) for tgt in transitive_targets.dependencies
+            ],
+            for_sources_types=(FileSourceField, ResourceSourceField),
+            # enable_codegen=True,
+        ),
+    )
+
+    source_files_get = Get(
+        SourceFiles,
+        SourceFilesRequest(field_set.source for field_set in request.field_sets),
+    )
+
+    # actually validate it with an external script.
+    # Validation cannot be inlined here because it needs to import the st2 code.
+    pex_get = Get(
+        VenvPex,
+        PexFromTargetsRequest(
+            [
+                Address(
+                    CMD_DIR,
+                    target_name="cmd",
+                    relative_file_path=f"{VALIDATE_CMD}.py",
+                ),
+            ],
+            output_filename=f"{VALIDATE_CMD}.pex",
+            internal_only=True,
+            main=EntryPoint.parse(f"{CMD_MODULE}.{VALIDATE_CMD}:main"),
+        ),
+    )
+
+    pex, dependency_files, source_files = await MultiGet(
+        pex_get, dependency_files_get, source_files_get
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests((dependency_files.snapshot.digest, source_files.snapshot.digest)),
+    )
+
+    process_result = await Get(
+        FallibleProcessResult,
+        VenvPexProcess(
+            pex,
+            argv=(
+                "--config-file",
+                "conf/st2.dev.conf",
+                # "--validate-defs",  # check for x-api-model in definitions
```

Comment:
Do we want to use the commented out values? Or are these just to show extra options we could use?

Reply:
Oh. I'll add an explanatory comment. These are things we might want to add in the future. We can't use `--validate-defs` yet, because there are a lot (hundreds?) of failures right now. The Makefile also does not run this yet.